extmod/network_wiznet5k_lwip.c

```diff
@@ -0,0 +1,570 @@
+#include "extmod/machine_spi.h"
+#include "extmod/virtpin.h"
+#include "modmachine.h"
+
+#if MICROPY_PY_WIZNET5K && MICROPY_PY_LWIP
+
+#include "shared/netutils/netutils.h"
+#include "lib/wiznet5k/Ethernet/wizchip_conf.h"
+#include "lib/wiznet5k/Ethernet/socket.h"
+#include "lwip/err.h"
+#include "lwip/dns.h"
+#include "lwip/dhcp.h"
+#include "netif/etharp.h"
+
+#define TRACE_ETH_TX (0x0002)
+#define TRACE_ETH_RX (0x0004)
+
+#ifndef MICROPY_HW_WIZNET_SPI_BAUDRATE
+#define MICROPY_HW_WIZNET_SPI_BAUDRATE  (4000000)
+#endif
+
+#ifndef WIZCHIP_SREG_ADDR
+#if (_WIZCHIP_ == 5500)
+#define WIZCHIP_SREG_ADDR(sn, addr)    (_W5500_IO_BASE_ + (addr << 8) + (WIZCHIP_SREG_BLOCK(sn) << 3))
+#else
+#define WIZCHIP_SREG_ADDR(sn, addr)    (_WIZCHIP_IO_BASE_ + WIZCHIP_SREG_BLOCK(sn) + (addr))
+#endif
+#endif
+
+#ifndef printf
+#define printf(...) mp_printf(MP_PYTHON_PRINTER, __VA_ARGS__)
+#endif
+
+/*******************************************************************************/
+// Wiznet5k Ethernet driver in MACRAW mode
+
+typedef struct _wiznet5k_obj_t {
+    mp_obj_base_t base;
+    mp_uint_t cris_state;
+    struct _machine_spi_obj_t *spi;
+    mp_hal_pin_obj_t cs;
+    mp_hal_pin_obj_t rst;
+    mp_hal_pin_obj_t pin_intn;
+    bool use_interrupt;
+    uint8_t eth_frame[1514];
+    uint32_t trace_flags;
+    struct netif netif;
+    struct dhcp dhcp_struct;
+} wiznet5k_obj_t;
+
+// Global object holding the Wiznet5k state
+STATIC wiznet5k_obj_t wiznet5k_obj;
+
+extern struct _machine_spi_obj_t *spi_from_mp_obj(mp_obj_t o);
+extern void wiznet5k_try_poll(void);
+
+STATIC void wiznet5k_lwip_init(wiznet5k_obj_t *self);
+
+STATIC void wiz_cris_enter(void) {
+    wiznet5k_obj.cris_state = MICROPY_BEGIN_ATOMIC_SECTION();
+}
+
+STATIC void wiz_cris_exit(void) {
+    MICROPY_END_ATOMIC_SECTION(wiznet5k_obj.cris_state);
+}
+
+STATIC void wiz_cs_select(void) {
+    mp_hal_pin_low(wiznet5k_obj.cs);
+}
+
+STATIC void wiz_cs_deselect(void) {
+    mp_hal_pin_high(wiznet5k_obj.cs);
+}
+
+void mpy_wiznet_yield(void) {
+    // Used in socket.c via -DWIZCHIP_YIELD=mpy_wiznet_yield in make/cmake
+    #if MICROPY_PY_THREAD
+    MICROPY_THREAD_YIELD();
+    #else
+    mp_handle_pending(true);
+    #endif
+}
+
+STATIC mp_obj_t mpy_wiznet_read_int(mp_obj_t none_in) {
+    (void)none_in;
+    wizchip_clrinterrupt(IK_SOCK_0);
+    #if _WIZCHIP_ == W5100S
+    setSn_IR(0, Sn_IR_RECV);
+    #endif
+
+    // Handle incoming data
+    wiznet5k_try_poll();
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_1(mpy_wiznet_read_int_obj, mpy_wiznet_read_int);
+
+STATIC void wiznet5k_config_interrupt(bool enabled) {
+    if (!wiznet5k_obj.use_interrupt) {
+        return;
+    }
+    mp_hal_pin_interrupt(
+        wiznet5k_obj.pin_intn,
+        MP_OBJ_FROM_PTR(&mpy_wiznet_read_int_obj),
+        (enabled)? MP_HAL_PIN_TRIGGER_FALL : MP_HAL_PIN_TRIGGER_NONE,
+        true
+        );
+}
+
+STATIC void wiz_spi_read(uint8_t *buf, uint16_t len) {
+    uint8_t *tx_buf = alloca(len);
```

Comment:
you don't need to allocate a tx buffer, you can just use `buf` for both tx and rx